Refactored Tilt with AKS docs (#5522)
kubernetes-sigs/cluster-api-provider-azure: docs/book/src/developers/tilt-with-aks-as-mgmt-ilb.md

- Teams that need to keep all Kubernetes API traffic within Azure VNet

**Note:** This is not a production ready solution and should not be used in a production environment. This is only meant to be used for development/testing purposes.

## Prerequisites

### Requirements
<!-- markdown-link-check-disable-next-line -->
- A [Microsoft Azure account](https://azure.microsoft.com/)
  - Note: If using a new subscription, make sure to [register](https://learn.microsoft.com/azure/azure-resource-manager/management/resource-providers-and-types) the following resource providers:
    - `Microsoft.Compute`
    - `Microsoft.Network`
    - `Microsoft.ContainerService`
    - `Microsoft.ManagedIdentity`
    - `Microsoft.Authorization`
    - `Microsoft.ResourceHealth` (if the `EXP_AKS_RESOURCE_HEALTH` feature flag is enabled)
- Install the [Azure CLI](https://learn.microsoft.com/cli/azure/install-azure-cli?view=azure-cli-latest)
- A [supported version](https://github.com/kubernetes-sigs/cluster-api-provider-azure#compatibility) of [clusterctl](https://cluster-api.sigs.k8s.io/user/quick-start#install-clusterctl)
- Basic understanding of Azure networking concepts, Cluster API, and CAPZ.

> - Go, Wget, and Tilt installed on your development machine.

Reviewer: ```suggestion
- `go`, `wget`, and `tilt` installed on your development machine.
```

- If `tilt-settings.yaml` file exists in the root of your repo, clear out any values in `kustomize_settings` unless you want to use them instead of the values that will be set by running `make aks-create`.

### Managed Identity & Registry Setup
1. Have a managed identity created from Azure Portal.
2. Add the following lines to your shell config such as `~/.bashrc` or `~/.zshrc`

>     ```

Reviewer 2: ```suggestion
   ```shell
```

>    export USER_IDENTITY="<Managed Identity name>"
>    export AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY="<your-client-id>"
>    export AZURE_CLIENT_ID="${AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY}"
>    export AZURE_OBJECT_ID_USER_ASSIGNED_IDENTITY="<your-object-id>"
>    export AZURE_LOCATION="<your-azure-location>"
>    export REGISTRY=<your-container-registry>

Reviewer: ```suggestion
   export USER_IDENTITY="<user-assigned-managed-identity-name>"
   export AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY="<user-assigned-managed-identity-client-id>"
   export AZURE_CLIENT_ID="${AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY}"
   export AZURE_OBJECT_ID_USER_ASSIGNED_IDENTITY="<user-assigned-managed-identity-object-id>"
   export AZURE_LOCATION="<azure-location-having-quota-for-B2s-and-D4s_v3-SKU>"
   export REGISTRY=<your-container-registry>
```

## Steps to Use Tilt with AKS as the Management Cluster

1. In tilt-settings.yaml, set subscription_type to "corporate" and remove any other env values unless you want to override env variables created by `make aks-create`. Example:
   ```
   .
   .
   .
   kustomize_substitutions:
   SUBSCRIPTION_TYPE: "corporate"
   .
   ```
2. `make clean`
   - This make target does not need to be run every time. Run it to remove bin and kubeconfigs.
3. `make generate`
   - This make target does not need to be run every time. Run it to update your go related targets, manifests, flavors, e2e-templates and addons.
   - Run it periodically upon updating your branch or if you want to make changes in your templates.
4. `make acr-login`
   - Run this make target only if you have `REGISTRY` set to an Azure Container Registry. If you used DockerHub like we recommend, you can skip this step.
5. `make aks-create`
   - Run this target to bring up an AKS cluster.
   - Once the AKS cluster is created, you can reuse the cluster as many times as you like. Tilt ensures that the new image gets deployed every time there are changes in the Tiltfile and/or dependent files.

>    - Running `make aks-create` cleans up any existing `aks_as_mgmt_settings`.

Reviewer: ```suggestion
   - Running `make aks-create` cleans up any existing variables from `aks_as_mgmt_settings` from the `tilt-settings.yaml`.
```

6. `make tilt-up`
   - Run this target to use underlying cluster being pointed by your `KUBECONFIG`.

7. Once the tilt UI is up and running click on the `allow required ports on mgmt cluster` task (checkmark the box and reload) to allow the required ports on the management cluster's API server.
   - Note: This task will wait for the NSG rules to be created and then update them to allow the required ports.
   - This task will take a few minutes to complete. Wait for this to finish to avoid race conditions.

> 8.  Checkmark the flavors you want to deploy and CAPZ will deploy the workload cluster with the selected flavor.

Reviewer 2: ```suggestion
8. Check the flavors you want to deploy and CAPZ will deploy the workload cluster with the selected flavor.
```

      - Flavors that leverage internal load balancer and are available for development in CAPZ for MSFT Tenant:

>          - [apiserver-ilb](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-apiserver-ilb.yaml): VM based default flavor that brings up native K8s clusters with Linux nodes.

Reviewer 2: ```suggestion
         - [apiserver-ilb](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-apiserver-ilb.yaml): VM-based default flavor that brings up native K8s clusters with Linux nodes.
```

>          - [apiserver-ilb-windows](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-windows-apiserver-ilb.yaml): VM based flavor that brings up native K8s clusters with Linux and Windows nodes.

Reviewer 2: ```suggestion
         - [apiserver-ilb-windows](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-windows-apiserver-ilb.yaml): VM-based flavor that brings up native K8s clusters with Linux and Windows nodes.
```

- Developers working in environments with strict network security requirements.
- Teams that need to keep all Kubernetes API traffic within Azure VNet

**Note:** This is not a production ready solution and should not be used in a production environment. This is only meant to be used for development/testing purposes.

## Prerequisites

### Requirements
<!-- markdown-link-check-disable-next-line -->
- A [Microsoft Azure account](https://azure.microsoft.com/)
  - Note: If using a new subscription, make sure to [register](https://learn.microsoft.com/azure/azure-resource-manager/management/resource-providers-and-types) the following resource providers:
    - `Microsoft.Compute`
    - `Microsoft.Network`
    - `Microsoft.ContainerService`
    - `Microsoft.ManagedIdentity`
    - `Microsoft.Authorization`
    - `Microsoft.ResourceHealth` (if the `EXP_AKS_RESOURCE_HEALTH` feature flag is enabled)
- Install the [Azure CLI](https://learn.microsoft.com/cli/azure/install-azure-cli?view=azure-cli-latest)

> - A [supported version](https://github.com/kubernetes-sigs/cluster-api-provider-azure#compatibility) of `clusterctl`

Reviewer 3: Can you link `clusterctl` to https://cluster-api.sigs.k8s.io/user/quick-start#install-clusterctl ?

Author: Did this but also kept the original link for supported version. Not sure if you would rather remove supported version link and have only clusterctl link.

- Basic understanding of Azure networking concepts, Cluster API, and CAPZ.
- Go, Wget, and Tilt installed on your development machine.
- If `tilt-settings.yaml` file exists in the root of your repo, clear out any values in `kustomize_settings` unless you want to use them instead of the values that will be set by running `make aks-create`.

### Managed Identity & Registry Setup
1. Have a managed identity created from Azure Portal.

> 2. Add the following lines to your shell config using `nano ~/.bashrc` or `nano ~/.zshrc`

Reviewer 3: ```suggestion
2. Add the following lines to your shell config such as `~/.bashrc` or `~/.zshrc`
```

    ```
   export USER_IDENTITY="<Managed Identity name>"

>    export AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY="<Client ID seen in the portal>"

Reviewer 3: ```suggestion
   export AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY="<your-client-id>"
```

   export AZURE_CLIENT_ID="${AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY}"

>    export AZURE_OBJECT_ID_USER_ASSIGNED_IDENTITY="<Object ID seen in the portal>"

Reviewer 3: ```suggestion
   export AZURE_OBJECT_ID_USER_ASSIGNED_IDENTITY="<your-object-id>"
```

>    export AZURE_LOCATION="<location of your choice>"

Reviewer 3: ```suggestion
   export AZURE_LOCATION="<your-azure-location>"
```

>    export REGISTRY=<registry of your choice>

Reviewer 3: ```suggestion
   export REGISTRY=<your-container-registry>
```

   ```

>    - We recommend using DockerHub: `export REGISTRY=docker.io/<DOCKERHUBUSERNAME>` then authenticate with `Docker login`.

Reviewer 3: ```suggestion
```

5. `make aks-create`
   - Run this target to bring up an AKS cluster.
   - Once the AKS cluster is created, you can reuse the cluster as many times as you like. Tilt ensures that the new image gets deployed every time there are changes in the Tiltfile and/or dependent files.
   - Running `make aks-create` cleans up any existing `aks_as_mgmt_settings`.
6. `make tilt-up`
   - Run this target to use underlying cluster being pointed by your `KUBECONFIG`.

7. Once the tilt UI is up and running click on the `allow required ports on mgmt cluster` task (checkmark the box and reload) to allow the required ports on the management cluster's API server.
   - Note: This task will wait for the NSG rules to be created and then update them to allow the required ports.
   - This task will take a few minutes to complete. Wait for this to finish to avoid race conditions.
8.  Checkmark the flavors you want to deploy and CAPZ will deploy the workload cluster with the selected flavor.
      - Flavors that leverage internal load balancer and are available for development in CAPZ for MSFT Tenant:

>          - [apiserver-ilb](../../../../templates/cluster-template-apiserver-ilb.yaml): VM based default flavor that brings up native K8s clusters with Linux nodes.

Reviewer 3: ```suggestion
         - [apiserver-ilb](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-apiserver-ilb.yaml): VM based default flavor that brings up native K8s clusters with Linux nodes.
```

>          - [apiserver-ilb-windows](../../../../templates/cluster-template-windows-apiserver-ilb.yaml): VM based flavor that brings up native K8s clusters with Linux and Windows nodes.

Reviewer 3: ```suggestion
         - [apiserver-ilb-windows](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-windows-apiserver-ilb.yaml): VM based flavor that brings up native K8s clusters with Linux and Windows nodes.
```
